src/physics/utilities/equation_solver.cpp

```diff
@@ -38,6 +38,169 @@ EquationSolver::EquationSolver(MPI_Comm comm, const LinearSolverParameters& lin_
+ * @brief A structure for storing simple JSON data
+ *
+ * Used to produce the configuration for AMGX
+ */
+class JSONTable {
+public:
+  /**
+   * @brief Creates a new table
+   * @param[in] depth The depth of the created table
+   */
+  JSONTable(const int depth = 0) : depth_(depth) {}
+  /**
+   * @brief Returns the subtable at the given index, creating a new table if necessary
+   * @param[in] idx The string key of the subtable (object)
+   */
+  JSONTable& operator[](const std::string& idx)
+  {
+    if (tables_.count(idx) == 0) {
+      tables_[idx] = std::make_unique<JSONTable>(depth_ + 1);
+    }
+    return *(tables_[idx]);
+  }
+
+  /**
+   * @brief Returns the literal at the given index, creating a new one if it doesn't exist
+   * @param[in] idx The string key of the subtable (object)
+   */
+  JSONLiteral& literal(const std::string& idx) { return literals_[idx]; }
+
+  /**
+   * @brief Adds (appends) a table to the calling table
+   * @param[in] other The table to add
+   */
+  void add(const JSONTable& other)
+  {
+    literals_.insert(other.literals_.begin(), other.literals_.end());
+    for (const auto& [key, subtable] : other.tables_) {
+      operator[](key).add(*subtable);
+    }
+  }
+
+  /**
+   * @brief Inserts the JSON representation to a stream
+   * @param[inout] out The stream to insert into
+   * @param[in] table The table to be inserted (printed)
+   */
+  friend std::ostream& operator<<(std::ostream& out, const JSONTable& table);
+
+private:
+  /**
+   * @brief Literal (primitive) members of the table
+   */
+  JSONLiterals literals_;
+  /**
+   * @brief Subtable (object) members of the table
+   */
+  std::unordered_map<std::string, std::unique_ptr<JSONTable>> tables_;
+  /**
+   * @brief Current depth of the table relative to the root
+   */
+  const int depth_;
+};
+
+std::ostream& operator<<(std::ostream& out, const JSONTable& table)
+{
+  out << "{";
+  std::string indent(table.depth_ * 2, ' ');  // Double-space indenting
+  char        sep = ' ';                      // Start with empty separator to avoid trailing comma
+  for (const auto& [key, val] : table.literals_) {
+    out << sep << "\n" << indent << "\"" << key << "\": ";
+    // Strings need to be wrapped in escaped quotes
+    if (auto str_ptr = std::get_if<std::string>(&val)) {
+      out << "\"" << *str_ptr << "\"";
+    } else {
+      std::visit([&out](const auto contained_val) { out << contained_val; }, val);
+    }
+    sep = ',';
+  }
+
+  // Recursively insert subtables into the stream
+  for (const auto& [key, subtable] : table.tables_) {
+    out << sep << "\n" << indent << "\"" << key << "\": ";
+    out << *subtable;
+    sep = ',';
+  }
+  out << "\n" << indent << "}";
+  return out;
+}
+
+std::unique_ptr<mfem::AmgXSolver> configureAMGX(const MPI_Comm comm, const AMGXPrec& options)
+{
+  auto              amgx                 = std::make_unique<mfem::AmgXSolver>();
+  static const auto default_prec_options = []() {
+    JSONTable result;
```

Comment:
`result` is an odd name for the thing you aren't returning.

Reply:
Good catch, I had initially written this a little differently